accepted/2024/trusted-publishers-oidc-for-nuget-push.md

```diff
@@ -0,0 +1,750 @@
+| Codeberg     | 6           | 3      | 0                | 0.01        | No?                                                                                                                                                    |
+| Gitea        | 1           | 1      | 0                | 0           | No?                                                                                                                                                    |
+
+(considering only latest package versions metadata of packages published in the past year, our definition of "active
+packages")
+
+For this reason, we are focused on GitHub Actions as the first Trusted Publisher that NuGet.org supports.
+
+### Step 1. add a GitHub Actions trust policy to your NuGet.org account
+
+Somehow, NuGet.org must verify that an uploaded package is coming from the right person.
+
+Prior to Trusted Publishers, an **API key** acted as both the username and password to authenticate as the package owner
+during package upload. For web UI upload on NuGet.org, the user's logged in session (via browser cookies) authenticates
+the user. In lieu of these two older authentication methods, a package author must tell NuGet.org that they *trust*
+package uploads GitHub Actions coming from a specific **repository**. In addition to restrictions on which GitHub
+repository and workflow are applicable to the trust policy, API scoping will be allowed similar to API keys.
+
+In short, a GitHub Actions trust policy is restricted in two ways:
+
+1. **Which GitHub Actions repository and workflow is trusted?**
+2. **What NuGet API operations can be used inside the GitHub actions workflow?**
+
+#### Which GitHub Actions repository and workflow is trusted?
+
+The user will sign in to NuGet.org, navigate to a new "Trusted Publishers" account page, and configure a trusted GitHub
+repository using a form like this:
+
+![Trust policy form](meta/trusted-publishers-oidc-for-nuget-push/trust-policy-form.png)
+
+In the form, the package author would provide 3 critical pieces of information:
+
+1. The GitHub organization or username owning the repository.
+2. The GitHub repository name.
+3. Some filter on which GitHub Action workflow runs should be allowed to push packages.
+
+The filter on which workflow runs should be allowed to publish packages can potentially be very specific. GitHub Actions
+supplies a variety of metadata in the OIDC token available to workflow runs. Refer to [Understanding the OIDC
+token](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect#understanding-the-oidc-token)
+to see all of the properties.
+
+For the sake of simplicity, we will start with the following workflow filters:
+- Filtering for a specific *branch*, as shown in the UI above (branch filter)
```

Comment:
Will this also support tags? For example, when a tag is created `v1.2.3` and a workflow is kicked off automatically on tag creation to build that version and then publish it.

Reply:
Hmmm generally a tag trust policy should be possible if the workflow: https://docs.github.com/en/actions/security-for-github-actions/security-hardening-your-deployments/about-security-hardening-with-openid-connect#filtering-for-a-specific-tag

Question for you: would you want this feature? It seems like you would need to update the trust policy for every release as your version tag increments.

Reply:
I guess more what I'd be after is a policy that matches a tag with pattern (e.g. `:ref:refs/tags/v*`) rather than exact match.

My typical release process is to draft a release for a tag in GitHub that has a specific version (e.g. `v1.2.3`) with the release notes prepared etc. Then when I'm ready to ship, I publish that release which creates the tag. The triggers on my CI workflow are then setup like this:

```yml
on:
  push:
    branches: [ main ]
    tags: [ v* ]
  pull_request:
    branches: [ main ]
  workflow_dispatch:
```

Publishing then happens based on a condition like this:

```yaml
if: |
  github.event.repository.fork == false &&
  startsWith(github.ref, 'refs/tags/v')
```

The net effect is that the "real" build happens after tagging, and that packages get shipped off to NuGet.org once produced.

So in a vacuum, I would match on the tag rather than a branch as I wouldn't want _any_ push to main to be allowed to publish.

If tags aren't supported, then I would use an environment rather than a branch, as I currently use environments to scope the NuGet API key secret to that environment so it's not available to all the workflow runs (it also lets me manually approve the push job).

Reply:
I heartily endorse @martincostello's comments above - this is how _a very large_ set of OSS projects are versioned, not just in the .NET ecosystem but broadly across other ecosystems as well. Not supporting pattern-based, tag-based workflows would be a huge miss IMO.

Reply:
Thank you @martincostello and @baronfel for your input.

The full example helps a lot! Given branch, workflow file path, and environment do not have wildcard support, I think we should explicitly call out that tag filtering would support wildcard/pattern matching on the tag name. This is a reasonable idea and I will update the doc accordingly.

Reply:
Thanks. If it's useful to you, here are two full workflows (not 100% identical, as one has code signing) used to do NuGet publishing for [Polly](https://github.com/App-vNext/Polly/blob/main/.github/workflows/build.yml) and [one of my personal libraries](https://github.com/martincostello/xunit-logging/blob/main/.github/workflows/build.yml).

Reply:
It may be useful looking into branch wildcard support as well - I know that e.g. in dotnet we use pattern-based rules in many other places to signify some special status - e.g. for the SDK `^release/\d\.\d\.\d\d\d$` is a stable release branch and may be able to trigger certain workflows.

Reply:
I updated the document with wildcard support on tags and branches. This aligns with the pattern matching on package ID scopes. I am not sure if a regex is the right UX but I'm happy to hear other opinions on that!

Reply:
> I am not sure if a regex is the right UX but I'm happy to hear other opinions on that!

Glob patterns vs. regex have both been used as examples here, so I think it's worth clarifying which since the syntax is so similar for trivial cases.  A lot of places use glob patterns for branch matching, so there may be some confusion if regex is used.

Glob pattern above: `:ref:refs/tags/v*` (typically converted to regex: `:ref:refs/tags/v.*` though this is not precisely the same; it would be more accurate as `:ref:refs/tags/v[!/]*` though that's hardly intuitive)
Regex pattern above: `^release/\d\.\d\.\d\d\d$` (converted to glob: `release/[0-9].[0-9].[0-9][0-9][0-9]`)

Worth noting in the glob pattern example, typically glob expansion only happens within the current directory level, so `v*` will match `v1.2.3` but not `version/1.2.3`.  The simple regex conversion `.*` does not respect path delimiters, so `v.*` would match both `v1.2.3` and `version/1.2.3`.

Reply:
to clarify - I'd be perfectly happy for just globs, the regex was just meant to be a related example of branch and ref names being signals based on structure. I'd expect users to use `release/*` for example.

Reply:
Good point @jimmylewis, in my experience the term globbing is most often used for file system/file paths. I am not sure if the concept cleanly translated in users' minds to tag/branch names with slashes. IMO we can do something simple where `*` expands to `.*` regex and not worry about including slashes in that that capture.

Today "globbing" on NuGet.org API key scopes is like this:
https://github.com/NuGet/NuGetGallery/blob/a4b3c64acb65ca68eb67ac6b51242356aa5e4da7/src/NuGetGallery.Services/Extensions/ScopeExtensions.cs#L54-L58